### DISCUSSION TITLE
API generated classes are now abstract

### DISCUSSION THREAD

**Discussion by mike-novarca** [serverpod/serverpod#1857] [Q&A] [answered]
**Describe the bug**
The generated api class is now marked as abstract even though it is not an abstract class. This also breaks backwards compatibility thus making the version number not comply with the flutter standard.

**To Reproduce**
Declare an API in the yaml
Use Serverpod generate 
The generated class is now marked as abstract even though it is not an abstract class.

**Expected behavior**
Generated class not to be marked as abstract or it to be actually an abstract class and be properly documented with a use case.

**Screenshots**

**Serverpod versions**
 -  _CLI Version:_ Run `serverpod version` to determine the version of installed tooling.
 - _Library version:_ Check the pubspec.yaml file for the version of Serverpod you are using.

**Platform information**
On which platforms do you experience the issue. Please provide as much information here as you can. Where are you running the server, which front end device are you using?

**Additional context**

**Comment by Isakdl** [comment]
Could you please specify how this is breaking your code? 

The interface is still the same, you still have a constructor and you create the object in the exact same way as you did before. There is no functional change here. You should also be able to type check the object like before with like so:

```dart
var object = MyClass();
var isMyClass = object is MyClass; // true
```

The change to abstract is just an implementation detail.

**Comment by mike-novarca** [comment]
You cannot extend an abstract class, in this it is materially different from before.

**Comment by Isakdl** [comment, marked as answer]
Yes you are right in that we have broken the extend functionality of classes, it is a breaking change we missed. We tried hard not to create any breaking changes so this one slipped past unnoticed, (it breaks not because of the class being abstract, but because we use a factory constructor). 

I'm still curious about what your use case is, and if that is now impossible to achieve.

If you only need to add additional methods on the classes the recommended Serverpod solution is to use an extension on the class rather than subclassing.

**Comment by mike-novarca** [comment]
The workaround used it to create a new class that implements the abstract class, copying in all the methods from the generated code. This is far from ideal as if the generated code changes it will not be updated on the implementing class.

**Comment by vlidholt** [comment]
I moved this to a discussion as this is not something we are going to change. It's not ideal that it is a breaking change, but the intended way you extend the generated classes is to use a Dart `extension`. Otherwise, the use case is still very limited as serialization will not work correctly with sub classes to the generated code (even with the old setup).

**Comment by kieran-myopnotes** [comment]
But this is surely not an abstract class because there are implementations in there? 

Our use case is as follows: 
We have a variety of input types within the system (think of a notion style block editor) all consisting of the same base data structure yet varying dramatically in front-end implementation. We have created a protocol file called "InputType" which consists of these base parameters that all our "inputTypes" share. This was done so that we can easily send the data too and from the server and front-end without having to fill our protocol files with hundreds of different input types all with the same set of parameters.

We then create individual classes on the front-end for each input type that all extend this base "InputType" class taking care of all the required data and serialisation. We have then also created an interface that these classes also extend forcing all of them to implement a standard set of methods. The result is that whenever we want to create a new input type we simply create a new class on the front-end, extend the universal inputType class, implement the universal inputTypeInterface, and then write our methods such as the input types build method that returns a widget on how the input type should look.

After this change, we have now had to create an InputTypeBase class and copy all the stuff from the generated class just so that we can extend it which just feels a bit unnecessary
